File: client/src/features/book/types/book.ts
import type { Author } from '@/features/author/types/author.js';
import type { GenreName } from '@/features/genre/types/genre.js';
import type { ReadingStatusAPI } from '@/features/reading-list/types/readingList.js';
import type { Review } from '@/features/review/types/review.js';

export interface Book {
  id: string;
  title: string;
  authorName?: string | null;
  imagePath: string;
  shortDescription: string;
  averageRating: number;
  genres: GenreName[];
}

export interface CreateBook {
  title: string;
  authorId?: string | null;
  image?: File | null;
  shortDescription: string;
  longDescription: string;
  publishedDate?: string | null | undefined;
  genres: string[];
}

export interface BookDetails extends Book {
  publishedDate?: string | null;
  ratingsCount: number;
  longDescription: string;
  creatorId?: string | null;
  moreThanFiveReviews: boolean;
  isApproved: boolean;
  readingStatus?: ReadingStatusAPI | null;
  author?: Author | null;
  reviews: Review[];
}

File: server/BookHub/Features/Book/Data/Models/Book.cs
namespace BookHub.Features.Book.Data.Models
{
    using System.ComponentModel.DataAnnotations;
    using System.ComponentModel.DataAnnotations.Schema;
    using BookHub.Data.Models.Base;
    using BookHub.Data.Models.Shared.BookGenre;
    using Features.Authors.Data.Models;
    using Features.Identity.Data.Models;
    using Features.ReadingList.Data.Models;
    using Features.Review.Data.Models;

    using static Shared.ValidationConstants;

    public class Book : DeletableEntity<int>, IApprovableEntity
    {
        [Required]
        [MaxLength(TitleMaxLength)]
        public string Title { get; set; } = null!;

        [Required]
        [MaxLength(ShortDescriptionMaxLength)]
        public string ShortDescription { get; set; } = null!;

        [Required]
        [MaxLength(LongDescriptionMaxLength)]
        public string LongDescription { get; set; } = null!;

        public double AverageRating { get; set; }

        public int RatingsCount { get; set; }

        public DateTime? PublishedDate { get; set; }

        [Required]
        [MaxLength(ImageUrlMaxLength)]
        public string ImageUrl { get; set; } = null!;

        [ForeignKey(nameof(Author))]
        public int? AuthorId { get; set; }

        public AuthorDbModel? Author { get; set; }

        [ForeignKey(nameof(User))]
        public string? CreatorId { get; set; }

        public User? Creator { get; set; }

        public bool IsApproved { get; set; }

        public ICollection<BookGenre> BooksGenres { get; } = new HashSet<BookGenre>();

        public ICollection<Review> Reviews { get; } = new HashSet<Review>();

        public ICollection<ReadingList> ReadingLists { get; } = new HashSet<ReadingList>();
    }
}
